engine/src/main/java/org/terasology/rendering/opengl/fbms/DynamicFBM.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.opengl.fbms;
+
+import java.nio.ByteBuffer;
+import org.lwjgl.BufferUtils;
+import org.lwjgl.opengl.Display;
+import org.lwjgl.opengl.GL11;
+import org.terasology.assets.ResourceUrn;
+import org.terasology.config.Config;
+import org.terasology.config.RenderingConfig;
+import org.terasology.registry.CoreRegistry;
+import org.terasology.rendering.oculusVr.OculusVrHelper;
+import org.terasology.rendering.opengl.AbstractFBM;
+import org.terasology.rendering.opengl.DefaultDynamicFBOs;
+import org.terasology.rendering.opengl.FBO;
+import org.terasology.rendering.opengl.FBOConfig;
+import org.terasology.rendering.opengl.PostProcessor;
+
+/**
+ * TODO: Add javadocs
+ * TODO: Better naming
+ */
+public class DynamicFBM extends AbstractFBM {
+    // I could have named them fullResolution, halfResolution and so on. But halfScale is actually
+    // -both- fullScale's dimensions halved, leading to -a quarter- of its resolution. Following
+    // this logic one32thScale would have to be named one1024thResolution and the otherwise
+    // straightforward connection between variable names and dimensions would have been lost. -- manu3d
+    private FBO.Dimensions fullScale;
+
+    private Config config = CoreRegistry.get(Config.class);
+    private RenderingConfig renderingConfig = config.getRendering();
+    private PostProcessor postProcessor;
+
+    @Override
+    public void initialise() {
+        super.initialise();
+        fullScale = new FBO.Dimensions(Display.getWidth(), Display.getHeight());
+        allocateFBO(DefaultDynamicFBOs.ReadOnlyGBuffer.getFboConfig());
+        allocateFBO(DefaultDynamicFBOs.WriteOnlyGBuffer.getFboConfig());
+        allocateFBO(DefaultDynamicFBOs.Final.getFboConfig());
+    }
+
+
+    /**
+     * Invoked before real-rendering starts
+     * TODO: how about completely removing this, and make Display observable and this FBM as an observer
+     */
+    public void update() {
+        updateFullScale();
+        if (getFBO(DefaultDynamicFBOs.ReadOnlyGBuffer.getResourceUrn()).dimensions().areDifferentFrom(fullScale)) {
+            disposeAllFBOs();
+            createFBOs();
+        }
+    }
+
+    private void disposeAllFBOs() {
+        FBO fbo;
+        for (ResourceUrn urn : fboConfigs.keySet()) {
+            fbo = fboLookup.get(urn);
+            if (fbo != null) {
```

Comment:
Hmmm, _defensive programming_. The reason why I choose to iterate fboConfigs is not to work around the [ConcurrentModificationException](https://docs.oracle.com/javase/7/docs/api/java/util/ConcurrentModificationException.html).